### PR TITLE
Fix #1690: Add support for setting submission limit per month per team for a challenge

### LINKED ISSUE #1690: Add support for setting submission limit  per month per team for a challenge
## Current Status

Currently, we only provide the capability to set the daily submission limit and total submission limit throughout the challenge. For evaluation of tasks envisioning evaluation of models on EvalAI as long-running test server and not challenge with finite time, we would like to add the feature to limit the number of submissions to a challenge per month per team which eventually limits the number of submissions to test-set per publication.

Thanks @jcjohnson for the feature request.

## Deliverables

- [ ] Add new field `max_submission_per_month` for each challenge phase
- [ ] Set default value to `max_submission_in_total`
- [ ] Add check for monthly submission limit 
- [ ] Add tests
- [ ] Update challenge creation API
- [ ] Modify API to get the remaining submissions and add support to show remaining monthly submissions
- [ ] Update yaml documentation
- [ ] Update EvalAI-Starters to add support for monthly max submission limit

### PR DESCRIPTION
**Pull request by vkartik97**
Closes #1690 <br>
**Deliverables**
- [x] Add new field max_submission_per_month for each challenge phase
- [x] Set default value to max_submission_in_total
- [x] Add check for monthly submission limit
- [x] Update challenge creation API
- [x] Modify API to get the remaining submissions and add support to show remaining monthly submissions
- [x] Add tests
- [x] Update yaml documentation
- [x] Update EvalAI-Starters to add support for monthly max submission limit~ [Fix EvalAI#1690:Updated to add support for monthly max submission limit](https://github.com/Cloud-CV/EvalAI-Starters/pull/3)
- [x] Handle case when daily submission limit and monthly submission limit gets exhausted at same time

### CODE REVIEW

**Comment by vkartik97** [comment]
@deshraj  @RishabhJain2018 
Please review

File: apps/challenges/models.py [hunk 1]
```diff
@@ -127,6 +127,7 @@ def __init__(self, *args, **kwargs):
     test_annotation = models.FileField(upload_to=RandomFileName("test_annotations"), default=False)
     max_submissions_per_day = models.PositiveIntegerField(default=100000, db_index=True)
     max_submissions = models.PositiveIntegerField(default=100000, db_index=True)
+    max_submissions_per_month = models.PositiveIntegerField(db_index=True)
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply apps/challenges/models.py]: Also, please add this line below line no. 128.
So, ideally the order should be:
1. max_submissions_per_day
2. max_submissions_per_month
3. max_submissions

File: apps/challenges/serializers.py [hunk 2]
```diff
@@ -141,6 +141,7 @@ class ChallengePhaseCreateSerializer(serializers.ModelSerializer):
 
     def __init__(self, *args, **kwargs):
         super(ChallengePhaseCreateSerializer, self).__init__(*args, **kwargs)
+        self.fields['max_submissions_per_month'].required = False
```
Comment (RishabhJain2018) [inline apps/challenges/serializers.py]: Please remove this condition as this field will be a required field.

File: apps/jobs/models.py [hunk 3]
```diff
@@ -146,6 +147,22 @@ def save(self, *args, **kwargs):
                     status=Submission.FAILED,
                     submitted_at__gte=datetime.date.today()).count()
 
+                submissions_in_month_count = Submission.objects.filter(
```
Comment (RishabhJain2018) [inline apps/jobs/models.py]: `submissions_in_month_count` -> `submissions_done_in_month_count`

File: apps/jobs/models.py [hunk 4]
```diff
@@ -146,6 +147,22 @@ def save(self, *args, **kwargs):
...
+                    challenge_phase__challenge=self.challenge_phase.challenge,
+                    participant_team=self.participant_team,
+                    challenge_phase=self.challenge_phase,
+                    submitted_at__gte=timezone.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0)).count()
+
+                failed_count_in_month = Submission.objects.filter(
```
Comment (RishabhJain2018) [inline apps/jobs/models.py]: `failed_count_in_month` -> `failed_submissions_count_in_month`

File: apps/jobs/models.py [hunk 5]
```diff
@@ -146,6 +147,22 @@ def save(self, *args, **kwargs):
...
+                    challenge_phase=self.challenge_phase,
+                    participant_team=self.participant_team,
+                    status=Submission.FAILED,
+                    submitted_at__gte=timezone.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0)).count()
+                if ((submissions_in_month_count + 1 - failed_count_in_month > self.challenge_phase.max_submissions_per_month) or
```
Comment (RishabhJain2018) [inline apps/jobs/models.py]: Can you please tell me why you are doing +1 here?

File: apps/jobs/views.py [hunk 6]
```diff
@@ -429,10 +431,38 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
     failed_submissions_done_today = submissions_done_today.filter(
         status=Submission.FAILED)
 
+    submissions_done_this_month = submissions_done.filter(
+        submitted_at__gte=timezone.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0))
+
+    failed_submissions_this_month = submissions_done_today.filter(
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: Why is there `submissions_done_today` ??

File: apps/jobs/views.py [hunk 7]
```diff
@@ -429,10 +431,38 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
+        current_month = date_time_now.month
+        # Get Next Month
+        year = date_time_now.year
+        next_month = date_time_now.month + 1
+        # Handling next_month for December
+        if next_month == 13 :
+            next_month = 1
+            year = date_time_now.year + 1
+        # Get Next Month's First Date
+        date_next_month = timezone.now().replace(year=year, month=next_month, day=1, hour=0, minute=0, second=0, microsecond=0)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: `date_next_month` -> `next_month_date`

File: apps/jobs/views.py [hunk 8]
```diff
@@ -429,10 +431,38 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
+        remaining_time = date_next_month - date_time_now
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: same as above!

File: apps/jobs/views.py [hunk 9]
```diff
@@ -429,10 +431,38 @@ def get_remaining_submissions(request, challenge_phase_pk, challenge_pk):
...
+        status=Submission.FAILED)
+
     submissions_done_count = submissions_done.count()
     failed_submissions_count = failed_submissions.count()
     submissions_done_today_count = submissions_done_today.count()
     failed_submissions_done_today_count = failed_submissions_done_today.count()
+    submissions_done_this_month_count = submissions_done_this_month.count()
+    failed_submissions_done_this_month_count = failed_submissions_this_month.count()
+
+    # Checks for Monthy Submission Limit
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: Ideally, it should come below the day's submission limit if case.

File: apps/challenges/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vkartik97) [reply apps/challenges/models.py]: >Please add default value of 100000.

To set default value to max_submission_in_total, as given in [deliverables](https://github.com/Cloud-CV/EvalAI/issues/1690)

File: apps/jobs/models.py (hunk 5 above)
Comment (vkartik97) [reply apps/jobs/models.py]: This will check if User's Current Submission [(`Successful Submission Made` + 1)<sup>th</sup>] is less than `max_submissions_per_month`.

File: apps/challenges/migrations/0038_challengephase_max_submissions_per_month.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (vkartik97) [reply apps/challenges/migrations/0038_challengephase_max_submissions_per_month.py]: @RishabhJain2018 
This is because there was Migration Conflict when I had [Merge branch 'master' into issue#1690](https://github.com/Cloud-CV/EvalAI/pull/1886/commits/872a34fb02636d104805a2fa4466269ba8a81c0d) and Commit to [Resolve Migration Conflicts](https://github.com/Cloud-CV/EvalAI/pull/1886/commits/ecc8213817be0a10f95042a41e5d56446b077a7d) because of addition of `0038_challenge_remote_evaluation.py`.

[the remark this answers is not in the document]
Comment (vkartik97) [reply apps/challenges/migrations/0038_challengephase_max_submissions_per_month.py]: Resolved

**Comment by vkartik97** [comment]
@RishabhJain2018 Please review

File: apps/jobs/views.py [hunk 11]
```diff
@@ -47,16 +47,16 @@
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: @vkartik97 Your text editor has done a lot of unnecessary changes. can you please revert all those.

**Comment by RishabhJain2018** [comment]
I believe that the daily submission limit will get exhausted first rather than monthly submission limit. If they both values are same and gets exhausted at the same time then we should show a message something like `Both daily and monthly submission limits are exhausted`.

**Comment by vkartik97** [comment]
@Ram81 
I have set `daily submission limit=monthly submission limit = maximum submissions=5` and I am getting everything Correctly:
screenshot from 2018-12-24 19-39-14
I tested this using Challenge Creation, Django admin and using PSQL.

**Comment by vkartik97** [comment]
@deshraj @RishabhJain2018 @Ram81 
If for a specific day, `submission_remaining_per_month < submission_remaining_today`, should I display `submission_remaining_per_month` for both `Today's Remaining Submissions` and `{Current_Month}'s Remaining Submissions`?

**Comment by vkartik97** [comment]
**Screenshot:**
When Both daily submission limit and monthly submission limit gets exhausted
screenshot from 2018-12-25 16-09-16

**Comment by vkartik97** [comment]
@RishabhJain2018 @deshraj @Ram81 Please review

**Comment by vkartik97** [comment]
Closed by mistake

**Comment by RishabhJain2018** [comment]
Closing this since #1994 is opened for the same!
